File: src/arp_parser.c
#include "arp_parser/arp_parser.h"

#include <arpa/inet.h>
#include <string.h>

int parse_arp(const uint8_t *data, size_t len, struct arp_packet *packet) {
  if (data == NULL || packet == NULL) return -1;
  if (len < 28) return -1;

  uint16_t htype = ntohs(*(uint16_t *)(data + 0));
  uint16_t ptype = ntohs(*(uint16_t *)(data + 2));
  uint8_t hsize = data[4];
  uint8_t psize = data[5];
  uint16_t op = ntohs(*(uint16_t *)(data + 6));

  if (hsize != 6 || psize != 4) return -2;
  if (htype != 1 || ptype != 0x0800) return -3;
  if (op < 1 || op > 2) return -4;

  packet->htype = htype;
  packet->ptype = ptype;
  packet->hsize = hsize;
  packet->psize = psize;
  packet->op = op;

  memcpy(packet->sha, data + 8, 6);
  memcpy(packet->spa, data + 14, 4);
  memcpy(packet->tha, data + 18, 6);
  memcpy(packet->tpa, data + 24, 4);

  // Семантические проверки
  //  мультикаст (224.0.0.0/4)
  if (packet->spa[0] >= 224 && packet->spa[0] <= 239) return -5;
  // Loopback (127.0.0.0/8)
  if (packet->spa[0] == 127) return -5;
  // ограниченный широковещательный адрес
  if (packet->spa[0] == 255 && packet->spa[1] == 255 && packet->spa[2] == 255 &&
      packet->spa[3] == 255)
    return -5;

  if (packet->tpa[0] >= 224 && packet->tpa[0] <= 239) return -6;
  if (packet->tpa[0] == 255 && packet->tpa[1] == 255 && packet->tpa[2] == 255 &&
      packet->tpa[3] == 255)
    return -6;

  return 0;
}

File: erlang/arp_parser.erl
-module(arp_parser).
-export([
    parse/1, 
    print_arp/1, 
    test/0
]).

-record(arp_packet, {
    htype,
    ptype,
    hsize,
    psize,
    op,
    sha,
    spa,
    tha,
    tpa
}).

parse(Binary) ->
    case Binary of
        <<HType:16, PType:16, HSize:8, PSize:8, Op:16, 
          SHa:6/binary, SPa:4/binary, THa:6/binary, TPa:4/binary>> ->
        #arp_packet{
            htype = HType, 
            ptype = PType, 
            hsize = HSize, 
            psize = PSize, 
            op = Op,
            sha = SHa, 
            spa = SPa, 
            tha = THa, 
            tpa = TPa
            };
        _ -> 
            {error, incomplete_packet}
    end.

print_arp(#arp_packet{
    htype = HType,
    ptype = PType,
    hsize = HSize,
    psize = PSize,
    op = Op,
    sha = SHa,
    spa = SPa,
    tha = THa,
    tpa = TPa
}) ->
    % Функция вывода record
    io:format("ARP Packet:~n"),
    io:format("  Hardware type: ~p~n", [HType]),
    io:format("  Protocol type: 0x~4.16.0B~n", [PType]),
    io:format("  Hardware size: ~p~n", [HSize]),
    io:format("  Protocol size: ~p~n", [PSize]),
    io:format("  Opcode: ~p~n", [Op]),
    io:format("  Sender MAC: ~s~n", [io_lib:format("~.16B:~.16B:~.16B:~.16B:~.16B:~.16B", binary_to_list(SHa))]),
    io:format("  Sender IP: ~p~n", [SPa]),
    io:format("  Target MAC: ~s~n", [io_lib:format("~.16B:~.16B:~.16B:~.16B:~.16B:~.16B", binary_to_list(THa))]),
    io:format("  Target IP: ~p~n", [TPa]),
    ok.

    test() ->
    TestPacket = <<
    16#00, 16#01, 16#08, 16#00, 16#06, 16#04, 16#00, 16#01,
    16#08, 16#00, 16#27, 16#12, 16#34, 16#56, 16#C0, 16#A8,
    16#01, 16#01, 16#00, 16#00, 16#00, 16#00, 16#00, 16#00,
    16#C0, 16#A8, 16#01, 16#02
    >>,
    case parse(TestPacket) of
        {error, Reason} -> io:format("error: ~p~n", [Reason]);
        Packet -> print_arp(Packet)
    end.